### PR TITLE
WIP: Tz bug fix rebased

### PR DESCRIPTION
**Pull request by lr4d**
Supersedes https://github.com/JDASoftwareGroup/kartothek/pull/260

I tried to rebase this branch on top of master but it was quite the adventure to do so.
I excluded the changes to the changelog and it seems like some removal of comments is also missing, but the important parts should be there.

cc @ged-steponavicius

### CODE REVIEW

File: kartothek/core/common_metadata.py [hunk 1]
```diff
@@ -78,6 +77,35 @@ def _schema_compat(self):
 
             schema = schema.remove_metadata()
             md = {b"pandas": _dict_to_binary(pandas_metadata)}
+            # https://github.com/JDASoftwareGroup/kartothek/issues/259
+
+            if schema is not None:
```
Comment (marco-neumann-jdas) [inline kartothek/core/common_metadata.py:82]: why can it be `None` here? 3 lines above we just call `schema = schema.remove_metadata()`. Can this return `None`?

File: kartothek/core/common_metadata.py [hunk 2]
```diff
@@ -78,6 +77,35 @@ def _schema_compat(self):
...
+                fields = []
+                for f in schema:
```
Comment (marco-neumann-jdas) [inline kartothek/core/common_metadata.py:84]: short comment before the entire massaging would be nice. This function just does more and more and it is hard to follow along.

File: kartothek/core/common_metadata.py [hunk 3]
```diff
@@ -380,24 +408,14 @@ def store_schema_metadata(schema, dataset_uuid, store, table):
...
-        f = schema[idx]
-
-        # schema data recovered from parquet always contains timestamp data in us-granularity, but pandas will use
-        # ns-granularity, so we re-align the two different worlds here
-        if f.type == pa.timestamp("us"):
-            f = pa.field(f.name, pa.timestamp("ns"))
-
-        fields.append(f)
-    return pa.schema(fields, schema.metadata)
+    return schema
```
Comment (marco-neumann-jdas) [inline kartothek/core/common_metadata.py:418]: a direct return from the line above would work now as well.

File: reference-data/arrow-compat/batch_generate_references.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Comment (marco-neumann-jdas) [inline reference-data/arrow-compat/batch_generate_references.sh:1]: thanks a lot for adding this tool!

File: tests/serialization/test_arrow_compat.py [hunk 5]
```diff
@@ -53,4 +52,7 @@ def test_arrow_compat(arrow_version, reference_store, mocker):
     if arrow_version in ("0.14.1", "0.15.0", "0.16.0") and not ARROW_LARGER_EQ_0141:
         orig = orig.astype({"null": float})
 
+    if LooseVersion(arrow_version) < "0.13.0":
```
Comment (marco-neumann-jdas) [inline tests/serialization/test_arrow_compat.py:55]: can you add a short comment on the reason? Is this because timezones cannot be preserved with these old versions?

Comment (lr4d) [reply tests/serialization/test_arrow_compat.py:55]: Because the changes from @ged-steponavicius fixes https://github.com/JDASoftwareGroup/kartothek/issues/259, enabling proper compatibilty of datetimes with timezones i.e. these columns, and we currently only support pyarrow >= 0.13.0

Comment (fjetter) [reply tests/serialization/test_arrow_compat.py:55]: Then why do we need to check `< "0.13.0":` That's dead code, isn't it?

**Comment by lr4d** [comment]
This PR now also includes the appropriate changelog entry and I added some commits to update the reference Parquet files.

@ged-steponavicius feel free to push to this branch if you'd like to address the review comments

File: tests/serialization/test_arrow_compat.py (hunk 5 above)
Comment (lr4d) [reply tests/serialization/test_arrow_compat.py:55]: `arrow_version` here refers to the arrow version with which the reference data file was created

Comment (lr4d) [reply tests/serialization/test_arrow_compat.py:55]:
```suggestion
    if LooseVersion(arrow_version) < "0.13.0":  # gh-259: compatibilty of datetimes with timezones only supported in kartothek versions with pyarrow >= 0.13.0
```

Comment (marco-neumann-jdas) [reply tests/serialization/test_arrow_compat.py:55]: suggestion looks good to me, thanks :)
